subsys/canbus/canopen/CO_driver.c

```diff
@@ -0,0 +1,470 @@
+		CONTAINER_OF(item, struct canopen_tx_work_container, work);
+	CO_CANmodule_t *CANmodule = container->CANmodule;
+	struct zcan_frame msg;
+	CO_CANtx_t *buffer;
+	int err;
+	u16_t i;
+
+	CO_LOCK_CAN_SEND();
+
+	for (i = 0; i < CANmodule->tx_size; i++) {
+		buffer = &CANmodule->tx_array[i];
+		if (buffer->bufferFull) {
+			msg.id_type = CAN_STANDARD_IDENTIFIER;
+			msg.std_id = buffer->ident;
+			msg.dlc = buffer->DLC;
+			msg.rtr = (buffer->rtr ? 1 : 0);
+			memcpy(msg.data, buffer->data, buffer->DLC);
+
+			err = can_send(CANmodule->dev, &msg, K_NO_WAIT,
+				       canopen_tx_isr_callback, CANmodule);
+			if (err == CAN_TIMEOUT) {
+				break;
+			} else if (err != CAN_TX_OK) {
+				LOG_ERR("failed to send CAN frame (err %d)",
+					err);
+				CO_errorReport(CANmodule->em,
+					       CO_EM_GENERIC_SOFTWARE_ERROR,
+					       CO_EMC_COMMUNICATION, 0);
+
+			}
+
+			buffer->bufferFull = false;
+		}
+	}
+
+	CO_UNLOCK_CAN_SEND();
+}
+
+void CO_CANsetConfigurationMode(void *CANdriverState)
+{
+	/* No operation */
+}
+
+void CO_CANsetNormalMode(CO_CANmodule_t *CANmodule)
+{
+	CANmodule->CANnormal = true;
+}
+
+CO_ReturnError_t CO_CANmodule_init(CO_CANmodule_t *CANmodule,
+				   void *CANdriverState,
+				   CO_CANrx_t rxArray[], u16_t rxSize,
+				   CO_CANtx_t txArray[], u16_t txSize,
+				   u16_t CANbitRate)
+{
+	u16_t i;
+	int err;
+
+	LOG_DBG("rxSize = %d, txSize = %d", rxSize, txSize);
+
+	if (!CANmodule || !rxArray || !txArray || !CANdriverState) {
+		LOG_ERR("failed to initialize CAN module");
+		return CO_ERROR_ILLEGAL_ARGUMENT;
+	}
+
+	if (rxSize > CONFIG_CAN_MAX_FILTER) {
+		LOG_ERR("insufficient number of concurrent CAN RX filters"
+			" (needs %d, %d available)", rxSize,
+			CONFIG_CAN_MAX_FILTER);
+		return CO_ERROR_OUT_OF_MEMORY;
+	} else if (rxSize < CONFIG_CAN_MAX_FILTER) {
+		LOG_INF("excessive number of concurrent CAN RX filters enabled"
+			" (needs %d, %d available)", rxSize,
+			CONFIG_CAN_MAX_FILTER);
+	}
+
+	canopen_detach_all_rx_filters(CANmodule);
+	canopen_tx_queue.CANmodule = CANmodule;
+
+	CANmodule->dev = CANdriverState;
+	CANmodule->rx_array = rxArray;
+	CANmodule->rx_size = rxSize;
+	CANmodule->tx_array = txArray;
+	CANmodule->tx_size = txSize;
+	CANmodule->CANnormal = false;
+	CANmodule->first_tx_msg = true;
+	CANmodule->errors = 0;
+	CANmodule->em = NULL;
+
+	for (i = 0U; i < rxSize; i++) {
+		rxArray[i].ident = 0U;
+		rxArray[i].pFunct = NULL;
+		rxArray[i].filter_id = CAN_NO_FREE_FILTER;
+	}
+
+	for (i = 0U; i < txSize; i++) {
+		txArray[i].bufferFull = false;
+	}
+
+	err = can_configure(CANmodule->dev, CAN_NORMAL_MODE, KHZ(CANbitRate));
+	if (err) {
+		LOG_ERR("failed to configure CAN interface (err %d)", err);
+		return CO_ERROR_ILLEGAL_ARGUMENT;
+	}
+
+	CANmodule->configured = true;
+
+	return CO_ERROR_NO;
+}
+
+void CO_CANmodule_disable(CO_CANmodule_t *CANmodule)
+{
+	int err;
+
+	if (!CANmodule || !CANmodule->dev) {
+		return;
+	}
+
+	canopen_detach_all_rx_filters(CANmodule);
```

Comment:
```C
for (i = 0U; i < rxSize; i++) {
	rxArray[i].filter_id = CAN_NO_FREE_FILTER;
}
```
?

Reply:
This is for initialisation.

Reply:
Ah, you are doing this in canopen_detach_all_rx_filters